runtime/lua/vim/lsp.lua

```diff
@@ -808,6 +808,72 @@ function lsp.client()
   error()
 end
 
+---@private
+---Resolve a workspace for a client
+---
+---@param markers string[] list of search patterns
+---@param bufnr number
+---@return table|nil list of `{ uri:string, name: string }`
+local function resolve_workspace(markers, bufnr)
+  if not markers then
+    return
+  end
+  bufnr = resolve_bufnr(bufnr)
+  local bufname = vim.api.nvim_buf_get_name(bufnr)
+  local buf_path = vim.fs.dirname(bufname)
+  local anchor =
+    vim.fs.find(markers, { path = buf_path, stop = vim.loop.os_homedir(), upward = true })[1]
+  if not anchor then
+    return
+  end
+  local root_dir = vim.fs.dirname(anchor)
+  if not root_dir then
+    return
+  end
+  local ws = {
+    name = root_dir,
+    uri = vim.uri_from_fname(root_dir),
+  }
+  return ws
+end
+
+---@private
+---Returns true if a workspace marker for a given client is found
+---
+---@param client table see |vim.lsp.client|
+---@param config table
+---@param bufnr number
+---@return boolean|nil
+local function should_reuse_client(client, config, bufnr)
+  if client.name ~= config.name then
+    return
+  end
+  bufnr = resolve_bufnr(bufnr)
+  local ws = resolve_workspace(config.workspace_markers, bufnr)
+  if not ws then
+    return
+  end
+  for _, folder in ipairs(client.workspace_folders or {}) do
+    if ws.uri == folder.uri then
+      local _ = log.debug() and log.debug('found existing workspace_folders for', ws.uri)
+      return true
+    end
+  end
+  local prompt_opts = {
+    prompt = string.format('add %s to current workspaces?\n[y]es or [n]o:\t', ws.uri),
+    default = 'n',
+  }
+  local should_add
+  vim.ui.input(prompt_opts, function(input)
+    if input == 'y' or input == 'Y' or input == 'yes' or input == 'Yes' then
+      should_add = lsp.add_workspace_folder(ws, { id = client.id })
+    end
+    -- TODO: is there another way to dismiss the prompt?
+    api.nvim_command('redraw')
+  end)
+  return should_add
```

Comment:
Just FYI: This would be broken if `vim.ui.input` would have a async implementation. 

But see my overall review comment

Reply:
The original logic existed in `vim.lsp.buf.add_workspace_folder`.

I just switched it from `input()` to `vim.ui.input`, but yeah I was wonder if it's a better idea to use `vim.ui.select` instead.